consensus/core/src/block_manager.rs

```diff
@@ -1,53 +1,397 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use crate::block::{BlockAPI, BlockRef, VerifiedBlock};
-use std::collections::HashSet;
+use crate::block::{Block, BlockAPI, BlockRef, VerifiedBlock};
+use crate::context::Context;
+use crate::dag_state::DagState;
+use crate::error::ConsensusResult;
+use parking_lot::RwLock;
+use std::collections::{BTreeMap, HashSet};
+use std::sync::Arc;
+
+struct SuspendedBlock {
+    block: VerifiedBlock,
+    missing_ancestors: HashSet<BlockRef>,
+}
+
+impl SuspendedBlock {
+    fn new(block: VerifiedBlock, missing_ancestors: HashSet<BlockRef>) -> Self {
+        Self {
+            block,
+            missing_ancestors,
+        }
+    }
+}
 
 /// Block manager suspends incoming blocks until they are connected to the existing graph,
 /// returning newly connected blocks.
 /// TODO: As it is possible to have Byzantine validators who produce Blocks without valid causal
 /// history we need to make sure that BlockManager takes care of that and avoid OOM (Out Of Memory)
 /// situations.
-#[allow(dead_code)]
 pub(crate) struct BlockManager {
-    // TODO: dummy implementation, just keep all the block references
-    accepted_blocks: HashSet<BlockRef>,
+    /// Keeps all the suspended blocks. A suspended block is a block that is missing part of its causal history and thus
+    /// can't be immediately processed. A block will remain in this map until all its causal history has been successfully
+    /// processed.
+    suspended_blocks: BTreeMap<BlockRef, SuspendedBlock>,
+    /// A map that keeps all the blocks that we are missing (keys) and the corresponding blocks that reference the missing blocks
+    /// as ancestors and need them to get unsuspended. It is possible for a missing dependency (key) to be a suspended block, so
+    /// the block has been already fetched but it self is still missing some of its ancestors to be processed.
+    missing_ancestors: BTreeMap<BlockRef, HashSet<BlockRef>>,
```

Comment:
Let's use `BTree{Set,Map}` by default. The number of keys and locality will probably make the BTree variants more efficient. And it will make iteration deterministic.